cabal-install/src/Distribution/Client/ProjectBuilding/UnpackedPackage.hs

```diff
@@ -678,7 +679,32 @@ buildAndInstallUnpackedPackage
           runConfigure
         PBBuildPhase{runBuild} -> do
           noticeProgress ProgressBuilding
-          runBuild
+          hookDir <- (</> ".cabal/hooks") <$> getEnv "HOME"
+          -- run preBuildHook. If it returns with 0, we assume the build was
+          -- successful. If not, run the build.
+          code <-
+            rawSystemExitCode
+              verbosity
+              (Just srcdir)
+              (hookDir </> "preBuildHook")
+              [ (unUnitId $ installedUnitId rpkg)
+              , (getSymbolicPath srcdir)
+              , (getSymbolicPath builddir)
+              ]
```

Comment:
This needs much more discussion.

1. relying on `PATH` to find hooks is very improper imo (I see you already changed it)
    - this should be in something like `~/.cabal/hooks` and the user should be able to configure the location via the cabal config
2. just like git, we should only execute hooks if they're marked as executable (windows is a special case, see how we did it [in stack](https://github.com/commercialhaskell/stack/blob/c0c7bfa64b9fb10eaec2b9160d65026b8e0e2e0a/src/Stack/Setup.hs#L1217-L1221) as an example)
3. we need a discussion what information we want to expose (e.g. env vars that are available to all hooks, like CABAL_VERSION or GHC_VERSION?)

Reply:
The `PATH` thing has been replaced with looking up ~/.cabal/hooks/.

Reply:
As per the [discussion ticket](https://github.com/haskell/cabal/issues/9892#issuecomment-2095018572), this now uses a project local `cabalHooks` directory.